module/zfs/zio.c

```diff
@@ -3928,13 +3943,28 @@ zio_ddt_write(zio_t *zio)
 		need_dvas -= parent_dvas;
 	}
 
+	if (is_ganged) {
+		zp->zp_dedup = B_FALSE;
+		BP_SET_DEDUP(bp, B_FALSE);
+		zio->io_pipeline = ZIO_WRITE_PIPELINE;
+		ddt_exit(ddt);
+		return (zio);
+	}
+
 	/*
 	 * We need to write. We will create a new write with the copies
 	 * property adjusted to match the number of DVAs we need to need to
 	 * grow the DDT entry by to satisfy the request.
 	 */
 	zio_prop_t czp = *zp;
-	czp.zp_copies = need_dvas;
+	if (have_dvas > 0) {
+		czp.zp_copies = need_dvas;
+		czp.zp_gang_copies = 0;
+	} else {
+		czp.zp_copies = need_dvas;
+		czp.zp_gang_copies = need_dvas +
+		    (zp->zp_gang_copies - zp->zp_copies);
+	}
```

Comment:
You don't need the `czp.zp_copies` assignments conditional.  And I am not getting the `czp.zp_gang_copies` logic, why there is not just `zp->zp_gang_copies`?

Reply:
That was because need_dvas isn't always equal to `zp_copies - have_dvas` in that case, because if there's an IO in flight, we subtract `parent_dvas` from it. However, if there is an IO in flight, we should just set `czp.zp_gang_copies` to zero also; we can't gang in that case either.

In fact we probably also need to have additional logic in that case to cause the new write to fail and fall back to a non-dedup write if the dde_lead_zio ends up ganging (or falling back in turn), since we can't add on to that.